zig: fix darwin SDK lookup

niklaskorz opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Building Zig packages on darwin can [currently fail](https://github.com/NixOS/nixpkgs/pull/373289#issuecomment-2607859906) because of the hardcoded `/usr/bin/xcrun` call in Zig's macOS SDK lookup method (https://github.com/ziglang/zig/issues/22600).

https://github.com/ziglang/zig/blob/f3c29dcb247be5e50d2d05f9fb2e814ebf115a86/lib/std/zig/system/darwin.zig#L51

This is fixed by replacing it with just `xcrun`, which will then use the `xcrun` provided by the nixpkgs darwin stdenv. I decided against using an absolute nix store path for `xcrun` as replacement as @reckenrode raised concerns about that causing issues in Nix development shells.

Zig also calls `/usr/bin/xcode-select`, which I have replaced by `xcode-select` in the `xcbuild` package. In this case, the absolute path to the nix store is acceptable even for devShells as Zig discards the output of `xcode-select` and just checks if the command is successful:

https://github.com/ziglang/zig/blob/f77e1b86225cd49c2d04dfa6ca4a7ede315dc0b1/lib/std/zig/system/darwin.zig#L18

~~I also removed the by now rather old Zig versions 0.9 and 0.10 (not completely arbitrarily, they are the only releases marked as "pre-release" on https://github.com/ziglang/zig/releases and had either no or only one dependant in nixpkgs). For `zig_0_9`, there was a single dependant [colorstorm](https://github.com/benbusby/colorstorm) which already is Zig 0.13 compatible in its HEAD state. I opened [an issue](https://github.com/benbusby/colorstorm/issues/16) requesting a new tagged release for colorstorm as the last one is three years old, and in the meanwhile bumped colorstorm to the commit bringing Zig 0.13 compatibility.~~ This part has been moved to a separate PR: #376540

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

niklaskorz commented:
Zig 0.11 still fails to build in the sandbox, looking into it

niklaskorz commented:
`pkgs/development/compilers/zig/generic.nix` appears to be completely unused since #314987 but I don't intend to address whether this file should be removed or all three remaining Zig versions can be combined into one package definition in _this_ PR. Instead, I just applied the darwin sandbox fixes to `generic.nix` too.

niklaskorz commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 376427`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 6 packages marked as broken and skipped:</summary>
  <ul>
    <li>cyber</li>
    <li>ghostty</li>
    <li>ghostty.man</li>
    <li>ghostty.shell_integration</li>
    <li>ghostty.terminfo</li>
    <li>ghostty.vim</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>cargo-lambda</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 26 packages built:</summary>
  <ul>
    <li>arocc</li>
    <li>aroccPackages.latest-unwrapped</li>
    <li>aroccStdenv</li>
    <li>cargo-zigbuild</li>
    <li>clipbuzz</li>
    <li>colorstorm</li>
    <li>dt</li>
    <li>emacsPackages.zig-mode</li>
    <li>findup</li>
    <li>flow-editor</li>
    <li>glsl_analyzer</li>
    <li>linuxwave</li>
    <li>minizign</li>
    <li>ncdu</li>
    <li>superhtml</li>
    <li>zf</li>
    <li>zig</li>
    <li>zig.doc</li>
    <li>zigStdenv</li>
    <li>zig_0_11</li>
    <li>zig_0_11.doc</li>
    <li>zig_0_12</li>
    <li>zig_0_12.doc</li>
    <li>zls</li>
    <li>zon2nix</li>
    <li>ztags</li>
  </ul>
</details>

Samasaur1 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 376427`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 6 packages marked as broken and skipped:</summary>
  <ul>
    <li>cyber</li>
    <li>ghostty</li>
    <li>ghostty.man</li>
    <li>ghostty.shell_integration</li>
    <li>ghostty.terminfo</li>
    <li>ghostty.vim</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 27 packages built:</summary>
  <ul>
    <li>arocc</li>
    <li>aroccPackages.latest-unwrapped</li>
    <li>aroccStdenv</li>
    <li>cargo-lambda</li>
    <li>cargo-zigbuild</li>
    <li>clipbuzz</li>
    <li>colorstorm</li>
    <li>dt</li>
    <li>emacsPackages.zig-mode</li>
    <li>findup</li>
    <li>flow-editor</li>
    <li>glsl_analyzer</li>
    <li>linuxwave</li>
    <li>minizign</li>
    <li>ncdu</li>
    <li>superhtml</li>
    <li>zf</li>
    <li>zig</li>
    <li>zig.doc</li>
    <li>zigStdenv</li>
    <li>zig_0_11</li>
    <li>zig_0_11.doc</li>
    <li>zig_0_12</li>
    <li>zig_0_12.doc</li>
    <li>zls</li>
    <li>zon2nix</li>
    <li>ztags</li>
  </ul>
</details>

Samasaur1 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 376427`

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 6 packages marked as broken and skipped:</summary>
  <ul>
    <li>cyber</li>
    <li>ghostty</li>
    <li>ghostty.man</li>
    <li>ghostty.shell_integration</li>
    <li>ghostty.terminfo</li>
    <li>ghostty.vim</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 27 packages built:</summary>
  <ul>
    <li>arocc</li>
    <li>aroccPackages.latest-unwrapped</li>
    <li>aroccStdenv</li>
    <li>cargo-lambda</li>
    <li>cargo-zigbuild</li>
    <li>clipbuzz</li>
    <li>colorstorm</li>
    <li>dt</li>
    <li>emacsPackages.zig-mode</li>
    <li>findup</li>
    <li>flow-editor</li>
    <li>glsl_analyzer</li>
    <li>linuxwave</li>
    <li>minizign</li>
    <li>ncdu</li>
    <li>superhtml</li>
    <li>zf</li>
    <li>zig</li>
    <li>zig.doc</li>
    <li>zigStdenv</li>
    <li>zig_0_11</li>
    <li>zig_0_11.doc</li>
    <li>zig_0_12</li>
    <li>zig_0_12.doc</li>
    <li>zls</li>
    <li>zon2nix</li>
    <li>ztags</li>
  </ul>
</details>

Eveeifyeve commented:
This fixes the ghostty on darwin incident see here for context: https://github.com/NixOS/nixpkgs/pull/369788

drupol commented:
Looks like the CI is broken

emilazy commented:
> This fixes the ghostty on darwin incident see here for context: #369788

This has absolutely nothing to do with that.

WeetHet commented on pkgs/development/compilers/zig/0.11/default.nix:
```diff
@@ -58,12 +59,22 @@ stdenv.mkDerivation (finalAttrs: {
+      substituteInPlace lib/std/zig/system/darwin.zig \
+        --replace-fail /usr/bin/xcrun xcrun \
+        --replace-fail /usr/bin/xcode-select '${lib.getExe' xcbuild "xcode-select"}'
+    '';
```
I don't like that this is duplicated three times over the span of three compilers. This is something unlikely to change in the future compilers as well, so this will continue to be duplicated in 0.14, 0.15, ...

WeetHet replied:
This also applies to stuff above, but I don't know how you can select a range of lines in GitHub PR reviews

niklaskorz replied:
This is already addressed by https://github.com/NixOS/nixpkgs/pull/331011

WeetHet replied:
Amazing, I didn't know about that PR, sorry :(

niklaskorz commented:
> Looks like the CI is broken

Zig 0.11 for Linux is broken on master independent of this PR:

- https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.zig_0_11.x86_64-linux
- https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.zig_0_11.aarch64-linux

niklaskorz commented:
Upstream already merged an according patch in https://github.com/ziglang/zig/pull/22611 so we can apply https://github.com/ziglang/zig/commit/92b20e42162675d35ffd27845e66b0f9213c00c2.patch for Zig <0.14 instead of doing the substitution in postPatch.

Samasaur1 commented:
> > > This fixes the ghostty on darwin incident see here for context: #369788
> > 
> > 
> > This has absolutely nothing to do with that.
> 
> Yes this does because `pkgs/apple-sdk` and `zig-objc` uses `getSdk` under the hood to get the apple sdk needed for ghostty on macos and we had the problem were ghostty at the end of the day was wrapping the dmg because we couldn't build on darwin this is where this pr solves the problem by fixing fetching apple sdks which makes it possible to build ghostty for source.

This will fix an issue that was encountered, but the larger problem that we're facing for building Ghostty from source is the Swift/xcodebuild situation, which this PR does not address

niklaskorz committed:
zig: fix darwin SDK lookup

niklaskorz commented:
Rebased on master and updated xcode-select substitution not to use a store path but just "xcode-select" so we do the same as the upstream fix at https://github.com/ziglang/zig/pull/22611, combined with xcbuild (which provides xcode-select) added to both zig's own native build inputs and to zig-hook's propagated inputs.

drupol commented on pkgs/development/compilers/zig/generic.nix:
```diff
@@ -92,16 +98,29 @@ stdenv.mkDerivation (finalAttrs: {
+      if lib.versionAtLeast finalAttrs.version "0.12" then
+        ''
+          substituteInPlace lib/std/zig/system.zig \
+            --replace-fail "/usr/bin/env" "${coreutils}/bin/env"
```
```suggestion
            --replace-fail "/usr/bin/env" "${lib.getExe' coreutils "env"}"
```

niklaskorz replied:
Addressed in 7d07c588d583491cbc8a3a112bd060db20bf9001

drupol commented on pkgs/development/compilers/zig/generic.nix:
```diff
@@ -92,16 +98,29 @@ stdenv.mkDerivation (finalAttrs: {
+      else
+        ''
+          substituteInPlace lib/std/zig/system/NativeTargetInfo.zig \
+            --replace-fail "/usr/bin/env" "${coreutils}/bin/env"
```
```suggestion
            --replace-fail "/usr/bin/env" "${lib.getExe' coreutils "env"}"
```

drupol reviewed (approved):
Diff LGTM, added a minor nit if you want to do it.

niklaskorz committed:
zig: refactor system env substitution

DimitarNestorov reviewed (approved):
<details>
<summary><code>cargo-lambda</code> failed for me but it also fails on master (I have sandbox enabled)</summary>

```

test test::test_download_example ... FAILED
test test::test_download_example_with_cache ... FAILED

failures:

---- test::test_download_example stdout ----
thread '<unnamed>' panicked at /private/tmp/nix-build-cargo-lambda-1.6.3.drv-1/cargo-lambda-1.6.3-vendor/hyper-0.14.32/src/server/server.rs:81:13:
error binding to 127.0.0.1:0: error creating server listener: Operation not permitted (os error 1)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'test::test_download_example' panicked at /private/tmp/nix-build-cargo-lambda-1.6.3.drv-1/cargo-lambda-1.6.3-vendor/httpmock-0.7.0/src/api/server.rs:301:37:
Cannot get server address: RecvError(())

---- test::test_download_example_with_cache stdout ----
thread '<unnamed>' panicked at /private/tmp/nix-build-cargo-lambda-1.6.3.drv-1/cargo-lambda-1.6.3-vendor/hyper-0.14.32/src/server/server.rs:81:13:
error binding to 127.0.0.1:0: error creating server listener: Operation not permitted (os error 1)
thread 'test::test_download_example_with_cache' panicked at /private/tmp/nix-build-cargo-lambda-1.6.3.drv-1/cargo-lambda-1.6.3-vendor/httpmock-0.7.0/src/api/server.rs:301:37:
Cannot get server address: RecvError(())


failures:
    test::test_download_example
    test::test_download_example_with_cache

test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass `-p cargo-lambda-invoke --lib`
note: keeping build directory '/private/tmp/nix-build-cargo-lambda-1.6.3.drv-1'
error: builder for '/nix/store/vrk236rzibq2lcs2apjil5g7vyc5a53f-cargo-lambda-1.6.3.drv' failed with exit code 101;
```
</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 6 packages marked as broken and skipped:</summary>
  <ul>
    <li>cyber</li>
    <li>ghostty</li>
    <li>ghostty.man</li>
    <li>ghostty.shell_integration</li>
    <li>ghostty.terminfo</li>
    <li>ghostty.vim</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>cargo-lambda</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 26 packages built:</summary>
  <ul>
    <li>arocc</li>
    <li>aroccPackages.latest-unwrapped</li>
    <li>aroccStdenv</li>
    <li>cargo-zigbuild</li>
    <li>clipbuzz</li>
    <li>colorstorm</li>
    <li>dt</li>
    <li>emacsPackages.zig-mode</li>
    <li>findup</li>
    <li>flow-editor</li>
    <li>glsl_analyzer</li>
    <li>linuxwave</li>
    <li>minizign</li>
    <li>ncdu</li>
    <li>superhtml</li>
    <li>zf</li>
    <li>zig</li>
    <li>zig.doc</li>
    <li>zigStdenv</li>
    <li>zig_0_11</li>
    <li>zig_0_11.doc</li>
    <li>zig_0_12</li>
    <li>zig_0_12.doc</li>
    <li>zls</li>
    <li>zon2nix</li>
    <li>ztags</li>
  </ul>
</details>

niklaskorz commented:
> ❌ 1 package failed to build:
> 
>     * cargo-lambda

That one is addressed in https://github.com/NixOS/nixpkgs/pull/379874
